docs/specs/forecast-api.md

```diff
@@ -176,16 +160,13 @@ Coordinates:
   * latitude   (latitude) float32 3kB -80.0 -79.75 -79.5 ... 79.5 79.75 80.0
   * flight_level      (flight_level) int16 2B 300
   * time       (time) datetime64[ns] 8B 2024-07-01T12:00:00
+    forecast_reference_time (time) datetime64[ns] 8B 2024-07-01T06:00:00
 Data variables:
     contrails   (longitude, latitude, flight_level, time) float32 4MB ...
 Attributes:
-    forecast_reference_time:  "2024-07-01T06:00:00Z"
```

Comment:
qq: What's the reason for migrating `forecast_reference_time` to be a coordinate and no longer an attribute? We have a similar attribute today and would then evaluate the same change!

Reply:
If you want to commit to only ever serving a single time slice, then I think having it as an attr makes sense. If you want to option available to return multiple time slices, then it should move to a coord (that references `time` 1:1).

--

The `forecast_reference_time` is the time at which the _meteorological model_ (used in generating the contrails predictions) was executed for a given `time`.  So multiple time slices may have different `forecast_reference_time` values.

Possible alternatives I can think of, but that seem less favorable:
- support multiple times in query but return multiple netcdfs for a given call, keeping `forecast_reference_time` as an attr
- support multiple times in query and concat slices into a single netcdf, but have all time slices reference a single `forecast_reference_time` (this would (1) limit the valid time slice candidates that could be concat'ed into a single file, and, (2) force choosing `time` slices that share a common `forecast_reference_time` instead of the latest available `forecast_reference_time` for a given `time`.

--
FWIW, I think its a good idea to stay away from supporting multiple time slices or time ranges in a query. It would make adding a 305 response option tricky in the future (many customers we've talked to expect to use the API as part of a caching service, so a 305 would be helpful).

Reply:
pinning this here as it is a relevant reminder for context ⏫ 

https://github.com/contrailcirrus/contrail-forecast/blob/refactor/v1-signature/docs/specs/forecast-api.md#timestamp

> Given that ECMWF delivers 73 hours of forecast data every 6 hours, it is expected that a given ts will have multiple candidate URIs to serve to a client.

> If multiple candidate URIs exist for a given timestamp at the time of the client request, the API should return the URI whose outputs were generated *using the nearest forecast time*.

Reply:
Thanks for elaborating, this really helped my understanding!

1) What makes caching a NetCDF response that has multiple concatenated time slices harder than caching a single time slice? Anticipating multiple time slices might otherwise be great API design! :)

2) Would it be backwards compatible for us to add the new reference time dimension to the existing NetCDF response, or is adding a new dimension generally considered breaking?

Reply:
> What makes caching a NetCDF response that has multiple concatenated time slices harder than caching a single time slice? Anticipating multiple time slices might otherwise be great API design! :)

TL;DR -- if we have N time slices, it is very possible that only a few slices are modified at any given time and warrant cache updates.  Sending N slices over the wire every time is sub-optimal.  Providing change detection (modification) on a per time slice basis, and enforcing retrieval on a per-time-slice basis, more strongly guarantees a client will have quality caching behavior (albeit with a bit more complexity to manage concurrency).

Longer answer:
If the contrail predictions are using ECMWF meteorological forecasts, then there will be (nominally) 12 different met forecasts for a given target `time`.  (ECMWF provides forecasts 73hrs into the future, now() inclusive, and they re-run the forecast every 6hrs, so any given hourly `time` has 73/6 overlapping forecast options).

As such, a query for a given `?time=splat` can return 12 different contrail forecasts, depending on the time of the call (default behavior, as per customer request and expected usage, being to return the latest forecast available).  This is why there is a note about the API not being idempotent. This is akin to checking for Friday's weather today, then checking for Friday's weather again tomorrow, and expecting the weather values to be different.  Noting that a future "super-user" upgrade could certainly be to accept a `?forecast_time=splat` param, which would guarantee idempotent behavior, but make client integration logic considerably more complex.

So to the question of caching logic wrt multiple time slices, a given netCDF file would have 12**N variants given N `time` slices in the netCDF.  A 305 response (or other similar change detection mechanism) would trigger whenever _any_ of the N slices is modified.  If a client is well informed wrt how the forecasts are due to update, they could get a high hit-rate of modified slides (N_mod/M_tot ~=1), but, making this assumption seems antithetical to a flexible API design (separating domain logic concerns from integration logic concerns). 

This is a bit sticky and I can see a few other ways that this could have been designed, but I won't elaborate here (this design was chosen given our opinions on how flight planners will want machine-2-machine integration without a big integration logic burden).  I'm happy to jump on a call and discuss the time/forecast_time behavior in more depth if you're interested.

> Would it be backwards compatible for us to add the new reference time dimension to the existing NetCDF response, or is adding a new dimension generally considered breaking?

Great question. I actually don't know from a netCDF consumer perspective. I'll inquire w/ our team today as they have more experience working w/ netCDF using various client-side tools. 

I _would_ expect that having the `forecast_time` as an attr, and then removing it (migrating to a coord) would be breaking.  So perhaps we keep it as a coord to start? (I don't see other substantive downside other than that `forecast_time` "feels" more like an attr than a first-class coord).

Reply:
following up here.
It shouldn't be breaking to add a new non-indexed coordinate to the netCDF at a future time.

It is important, however, to include `forcast_reference_time`... and if we include it in an attribute and then decide to migrate to a non-indexed coord later, it would be a breaking change to remove the _attribute_ in favor of the new non-indexed coordinate.

So if there is any possibility of wanting to return multiple `time` slices in the future, it seems reasonable to just have `forecast_reference_time` be a non-indexed coord (or, we can agree that we won't leave the back door open for aggregations on `time` and leave `forecast_reference_time` an attr... ◀️  my preference).